pkg/tools/socket.go

```diff
@@ -21,116 +22,151 @@ const (
-		}))
+func (b *builder) TCP() *builder {
+	return b.Network("tcp")
+}
 
-	return grpc.DialContext(ctx, addr.String(), opts...)
+func (b *builder) Unix() *builder {
+	return b.Network("unix")
 }
 
-// DialTimeout tries to establish connection with addr during timeout
-func DialTimeout(addr net.Addr, timeout time.Duration, opts ...grpc.DialOption) (*grpc.ClientConn, error) {
-	ctx, cancel := context.WithTimeout(context.Background(), timeout)
-	defer cancel()
-	return DialContext(ctx, addr, opts...)
+func (b *builder) Network(network string) *builder {
+	b.opts = append(b.opts, grpc.WithContextDialer(func(ctx context.Context, target string) (net.Conn, error) {
+		return (&net.Dialer{}).DialContext(ctx, network, target)
+	}))
+	return b
 }
 
-// DialContextUnix establish connection with passed unix socket
-func DialContextUnix(ctx context.Context, path string, opts ...grpc.DialOption) (*grpc.ClientConn, error) {
-	addr, err := net.ResolveUnixAddr("unix", path)
-	if err != nil {
-		return nil, err
-	}
-	return DialContext(ctx, addr, opts...)
+func (b *builder) Timeout(t time.Duration) *builder {
+	b.t = t
+	return b
 }
 
-// DialTimeoutUnix tries to establish connection with passed unix socket during timeout
-func DialTimeoutUnix(path string, timeout time.Duration, opts ...grpc.DialOption) (*grpc.ClientConn, error) {
-	ctx, cancel := context.WithTimeout(context.Background(), timeout)
-	defer cancel()
-	return DialContextUnix(ctx, path, opts...)
+func (b *builder) DialContextFunc() dialContextFunc {
+	return func(ctx context.Context, target string, opts ...grpc.DialOption) (conn *grpc.ClientConn, err error) {
+		if b.t != 0 {
+			var cancel context.CancelFunc
+			ctx, cancel = context.WithTimeout(ctx, b.t)
+			defer cancel()
+		}
+
+		if GetConfig().OpenTracing {
+			b.opts = append(b.opts, OpenTracingDialOptions()...)
+		}
+
+		if GetConfig().Insecure {
+			b.opts = append(b.opts, grpc.WithInsecure())
+		}
+
+		b.opts = append(b.opts, grpc.WithBlock())
+
+		return grpc.DialContext(ctx, target, append(opts, b.opts...)...)
+	}
 }
 
-// DialUnix simply calls DialTimeoutUnix with default timeout
-func DialUnix(path string, opts ...grpc.DialOption) (*grpc.ClientConn, error) {
-	return DialTimeoutUnix(path, dialTimeoutDefault, opts...)
+// OpenTracingDialOptions returns array of grpc.DialOption that should be passed to grpc.Dial to enable opentracing
+func OpenTracingDialOptions() []grpc.DialOption {
+	return []grpc.DialOption{
+		grpc.WithUnaryInterceptor(
+			otgrpc.OpenTracingClientInterceptor(opentracing.GlobalTracer(), otgrpc.LogPayloads())),
+		grpc.WithStreamInterceptor(
+			otgrpc.OpenTracingStreamClientInterceptor(opentracing.GlobalTracer())),
+	}
 }
 
-// DialContextTCP establish TCP connection with address
-func DialContextTCP(ctx context.Context, address string, opts ...grpc.DialOption) (*grpc.ClientConn, error) {
-	addr, err := net.ResolveTCPAddr("tcp", address)
+func readConfiguration() (DialConfig, error) {
```

Comment:
Maybe `readOpentracingConfiguration`?

Reply:
I suppose it will be not only opentracing but also security.